src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCache.cs

```diff
@@ -77,49 +78,76 @@ public CompilerCache(IFileProvider fileProvider)
                 throw new ArgumentNullException(nameof(compile));
             }
 
-            CompilerCacheResult cacheResult;
+            Task<CompilerCacheResult> cacheEntry;
             // Attempt to lookup the cache entry using the passed in path. This will succeed if the path is already
             // normalized and a cache entry exists.
-            if (!_cache.TryGetValue(relativePath, out cacheResult))
+            if (!_cache.TryGetValue(relativePath, out cacheEntry))
             {
                 var normalizedPath = GetNormalizedPath(relativePath);
-                if (!_cache.TryGetValue(normalizedPath, out cacheResult))
+                if (!_cache.TryGetValue(normalizedPath, out cacheEntry))
                 {
-                    cacheResult = CreateCacheEntry(normalizedPath, compile);
+                    cacheEntry = CreateCacheEntry(normalizedPath, compile);
                 }
             }
 
-            return cacheResult;
+            // The Task does not represent async work and is meant to provide per-entry locking.
+            // Hence it is ok to perform .GetResult() to read the result.
+            return cacheEntry.GetAwaiter().GetResult();
         }
 
-        private CompilerCacheResult CreateCacheEntry(
+        private Task<CompilerCacheResult> CreateCacheEntry(
             string normalizedPath,
             Func<RelativeFileInfo, CompilationResult> compile)
         {
-            var fileInfo = _fileProvider.GetFileInfo(normalizedPath);
-            MemoryCacheEntryOptions cacheEntryOptions;
-            CompilerCacheResult cacheResult;
-            if (!fileInfo.Exists)
+            TaskCompletionSource<CompilerCacheResult> compilationTaskSource = null;
+            RelativeFileInfo relativeFileInfo = null;
+            MemoryCacheEntryOptions cacheEntryOptions = null;
+            Task<CompilerCacheResult> cacheEntry;
+
+            lock (_cacheLock)
             {
-                var expirationToken = _fileProvider.Watch(normalizedPath);
-                cacheResult = new CompilerCacheResult(new[] { expirationToken });
+                if (_cache.TryGetValue(normalizedPath, out cacheEntry))
+                {
+                    return cacheEntry;
+                }
+
+                var fileInfo = _fileProvider.GetFileInfo(normalizedPath);
+                Task<CompilerCacheResult> cacheResult;
+                if (!fileInfo.Exists)
+                {
+                    var expirationToken = _fileProvider.Watch(normalizedPath);
+                    cacheResult = Task.FromResult(new CompilerCacheResult(new[] { expirationToken }));
+
+                    cacheEntryOptions = new MemoryCacheEntryOptions();
+                    cacheEntryOptions.AddExpirationToken(expirationToken);
+                }
+                else
+                {
+                    relativeFileInfo = new RelativeFileInfo(fileInfo, normalizedPath);
```

Comment:
Cant we construct this below and just save the `fileInfo`? That way it's being constructed by where it's used.